src/client.rs

```diff
@@ -172,7 +176,23 @@ impl Client {
             return Ok(Response::new(request.uri, Status::Excluded, request.source));
         }
         let status = match request.uri {
-            Uri::Website(ref url) => self.check_website(&url).await,
+            Uri::Website(ref url) => match self.check_website(url).await {
+                Status::Ok(code) if self.prefer_https && url.scheme() == "http" => {
+                    let mut https_url = url.clone();
+                    https_url
+                        .set_scheme("https")
+                        .unwrap_or_else(|_| unsafe { std::hint::unreachable_unchecked() });
```

Comment:
Not a fan of the `.unwrap_or_else(|_| unsafe { std::hint::unreachable_unchecked() })` part. I don't think it can fail, but can't we just use `?` instead?

Reply:
I cannot use `?` because it returns `Result<(), ()>`, where the error `()` cannot be turned into `StdError`.

I can use `unwrap` or `expect`, which carries extra codes. Or I can use `unreachable!`, but it doesn't allow optimization.

From https://doc.rust-lang.org/stable/std/hint/fn.unreachable_unchecked.html#safety
> Use this function only when you can prove that the code will never call it
which is exactly the case.

Reply:
Aha! Didn't know that. 😄 Would still prefer an expect here for a few reasons:
* It documents our expectations
* avoids using `unsafe`, which could be a requirement for projects which want to use lychee as a library
* If this statement could ever fail (e.g. after refactoring), at least we get an error message that could help us fix it

Reply:
I can understand the first and third points. But the fear of `unsafe` is unreasonable...

Reply:
There's another approach
```rust
debug_assert!(foo.is_ok(), "some error message");
foo.unwrap_or_else(|_| unsafe { std::hint::unreachable_unchecked() });
```

In debug build, you will get the same benefit of `expect`.
In release build, it allow optimization.

Reply:
> I can understand the first and third points. But the fear of unsafe is unreasonable...

Hum, you might be right. At the same time my intuition is to use unsafe only if necessary. In this case, what would be the advantage of using unsafe? Is the perf/size/compile difference thanks to the optimizations  measurable? At runtime, we'll mostly be bound by I/O I'd say.

The other approach is clever, but it would take me a while to understand what's going on; I might not even get it without a comment.
So, in summary I'd like to keep it simple and use an `expect` here unless I'm overlooking a big drawback. 🤔

Reply:
>  what would be the advantage of using unsafe? Is the perf/size/compile difference thanks to the optimizations measurable?

Yes.

> At runtime, we'll mostly be bound by I/O I'd say.

That's also true.

> The other approach is clever, but it would take me a while to understand what's going on; I might not even get it without a comment.

```rust
debug_assert!(foo.is_ok(), "some error message");
foo.unwrap_or_else(|_| unsafe { std::hint::unreachable_unchecked() });
```

This is simple to understand. `debug_assert` is noop in release build, and in debug build it acts like `assert`.

By asserting the variable of interest is a `Ok`, it's safe to unwrap it anyway.
Also, `assert` allows you to add custom error message.

In conclusion, it well solved the two points you mentioned above.
> - It documents our expectations
> - If this statement could ever fail (e.g. after refactoring), at least we get an error message that could help us fix it

Again, there's just no solid reason why to avoid `unsafe` here. It's mathematically impossible to err.

Reply:
Did some research. Performance gains are [debated](https://github.com/rust-lang/rust/issues/52945#issuecomment-409774203) in [this issue](https://github.com/rust-lang/rust/issues/52945) for the Rust compiler. It's arguably only faster in hot loops. We could do a benchmark to test that, but I'm not sure if it's worth it.

There's another comment by @oli-obk:

> I've been saved by the compiler panicking with "entered unreachable code" quite a few times. Especially during development where one might end up making such code reachable (maybe because a new features was implemented and some other code correctly assumed that it was unreachable before the existence of the new feature). Having these silently become UB seems... suboptimal.

He's a highly respected compiler dev, so I respect his opinion.
He mentions refactoring, which is also my argument. Even though the code is impossible to err, that might not be the case after refactoring. So, let's use expect.

Reply:
Nope, you're not listening.

The `debug_assert` test works even after refactoring.

Reply:
Put it in another way,

It an UB if it unwraps an Err. But the `debug_assert` explicitly check whether it's Err. If so, it panics with an error message, so that's what you can achieve with `expect`.

Then in release mode, it's OP, so no over head.

In one word, it's strictly better in every aspect than `expect`.

Reply:
Since I've been ~~summoned~~ pinged, I'm gonna throw in some unsolicited advice:

do benchmark the case using `unreachable_unchecked` against the unwrap case, and if you can really see a perf improvement here, do one of two things:

* debug asserts as suggested by @lebensterben 
* refactor the `set_scheme` function to return a `#[nonexhaustive]` empty enum (as the error type) instead of a `()` for now. You can add variants to that enum later if you decide that an error can actually happen. This way you get zero future API breakage (though you get it now), while the compiler can figure out that the `expect` call is dead code and eliminate it just as well as it would with the `unreachable_unchecked`.

If you don't see a perf improvement due to this `unreachable_unchecked`, either the compiler already figures it out for you, or this isn't really important. In either case my recommendation would be to stick with `expect`.

Reply:
Thanks for the input @oli-obk. Makes a lot of sense. @lebensterben, would you be interested in working on the benchmark?

Reply:
Later. I'm struggling with the side effect of COVID vaccine. I'm in and out of sleep all day.

Reply:
No rush. Hope you get well soon.